src/spmc_queue.ml

```diff
@@ -0,0 +1,244 @@
+   - Local dequeue and enqueue are quite close to wait freedom. Enqueuer can simply
+   check whether the slot it will get is free before incrementing tail, dequeuer can
+   just roll back after a few spins. I don't think it makes sense to implement that
+   in the general case though, as we don't actually want to return in those cases.
+   - Some accesses do not have to be atomic but Atomic.t does not have unsafe
+   methods. Probably doesn't matter on x86.
+*)
+
+type 'a t = {
+  head : int Atomic.t;
+  tail : int Atomic.t;
+  mask : int Atomic.t;
+  array : 'a option Atomic.t Array.t Atomic.t;
+}
+
+let create ~size_exponent () =
+  let size = Int.shift_left 1 size_exponent in
+  {
+    head = Atomic.make 0;
+    tail = Atomic.make 0;
+    mask = Atomic.make (size - 1);
+    array = Atomic.make (Array.init size (fun _ -> Atomic.make None));
+  }
+
+let indicative_size { head; tail; _ } =
+  max (Atomic.get tail - Atomic.get head) 0
+
+module Local = struct
+  (* [local_resize] resizes the array.
+
+     1. Create new array.
+     2. Overshoot the head to discourage any new stealers.
+     3. Transfer covered elements to the new array.
+     4. Ensure all other stealers finished (i.e. no elements in the old array).
+     5. Swap arrays.
+     6. Update mask, set tail and head to 0. In this order to discourage stealers.
+  *)
+  let resize t =
+    let { mask; array; tail; head; _ } = t in
+    let mask_val = Atomic.get mask in
+    let array_val = Atomic.get array in
+    let size = Atomic.get mask + 1 in
+    let new_size = size * 2 in
+    let new_mask = new_size - 1 in
+    let new_array_val = Array.init new_size (fun _ -> Atomic.make None) in
+    let old_head_val = Atomic.fetch_and_add head size in
+    let temp_head_val = old_head_val + size in
+    let tail_val = Atomic.get tail in
+    let num_of_items_to_copy = tail_val - old_head_val in
+    for i = 0 to num_of_items_to_copy - 1 do
+      let index_in_old_array = (old_head_val + i) land mask_val in
+      let old_cell = Array.get array_val index_in_old_array in
+      let value = Atomic.get old_cell in
+      (* Gc things, probably not necessary since old array should not live
+         for long. *)
+      Atomic.set old_cell None;
+      assert (Option.is_some value);
+      (* Why not start at index 0 in the new array? Easier than changing tail *)
+      let index_in_new_array = (old_head_val + i) land new_mask in
+      let new_cell = Array.get new_array_val index_in_new_array in
+      Atomic.set new_cell value
+    done;
+
+    (* just for extra safety, wait until stealers finish - should not
+       be necessary for correctness *)
+    let num_of_items_to_ensure_taken = temp_head_val - tail_val in
+    for i = 0 to num_of_items_to_ensure_taken - 1 do
+      let index = (tail_val + i) land mask_val in
+      while Option.is_some (Atomic.get (Array.get array_val index)) do
+        ()
+      done
+    done;
```

Comment:
This inner loop cannot be an assert, as stealers might still be removing items from the queue. 

But the conclusion is correct; both loops are not required for correctness. It's more for tractability of `steal`. Otherwise there's many ways to introduce subtle bugs, that we have no way of preventing, that is any access to mask, array, tail after CAS on head will break things.